File: java-solve/solve/src/s23_2/S23.java
package s23_2;

import structure.ListNode;

import java.util.PriorityQueue;

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Status> queue = new PriorityQueue<>();
        for (ListNode node : lists) {
            if (node != null) {
                queue.offer(new Status(node.val, node));
            }
        }
        ListNode head = new ListNode(0);
        ListNode tail = head;
        while (!queue.isEmpty()) {
            Status f = queue.poll();
            tail.next = f.ptr;
            tail = tail.next;
            if (f.ptr.next != null) {
                queue.offer(new Status(f.ptr.next.val, f.ptr.next));
            }
        }
        return head.next;
    }

    class Status implements Comparable<Status> {
        int val;
        ListNode ptr;

        Status(int val, ListNode ptr) {
            this.val = val;
            this.ptr = ptr;
        }

        public int compareTo(Status status2) {
            return this.val - status2.val;
        }
    }
}

public class S23 {
    public static void main(String[] args) {
        System.out.println(new Solution().mergeKLists(
                new ListNode[]{
                        ListNode.fromArray(new int[]{1, 4, 5}),
                        ListNode.fromArray(new int[]{1, 3, 4}),
                        ListNode.fromArray(new int[]{2, 6}),
                }
        ));
    }
}

File: python/heap/s23.py
import heapq
from typing import List

from structure import ListNode


class MyNode:
    def __init__(self, node: ListNode):
        self.node = node

    def __lt__(self, other):
        return self.node.val < other.node.val


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        my_lists=[MyNode(node) for node in lists]
        heapq.heapify(my_lists)
        dummy = ListNode(0)
        p = dummy
        while my_lists:
            node = heapq.heappop(my_lists).node
            p.next = node
            p = p.next
            if node.next:
                heapq.heappush(my_lists, MyNode(node.next))
        p.next = None
        return dummy.next
